apps/ewallet_db/lib/ewallet_db/auth_token.ex

```diff
@@ -100,19 +101,30 @@ defmodule EWalletDB.AuthToken do
   Generate an auth token for the specified user,
   then returns the auth token string.
   """
-  def generate(%User{} = user, owner_app, originator) when is_atom(owner_app) do
+  def generate(user, owner_app, originator, is_2fa_token \\ false)
+
+  def generate(%User{} = user, owner_app, originator, is_2fa_token) when is_atom(owner_app) do
     attrs = %{
       owner_app: Atom.to_string(owner_app),
       user_uuid: user.uuid,
       account_uuid: nil,
+      required_2fa: is_required_2fa(user, is_2fa_token),
```

Comment:
From the current flow in the PR description, a 2FA-enabled user will have an auth_token before 2FA is verified. A bit scary to me. 

Can we ask for 2FA token to be passed during `/admin.login` and the server returns an auth_token only when username/password/2FA all match?

---

What do you think? @ripzery @T-Dnzt @sirn

Reply:
@unnawut 

I like this way 👍 The current flow is 2-step verification, but it feels a bit scary for me too because having to manipulate the authentication token. (similar to go.exchange's flow)

I agree with passing the 2FA token to `/admin.login`. However, I have to choose the error response when the enabled-2fa user login with correct username/password but missing 2FA authentication token, what is a more appropriate error code between:
- `user:invalid_login_credentials`: 
 😃More secure by not giving a hint that which field is wrong.
 😟A bit less dynamic login flow.

- `user:some_2fa_error`:
 😃More dynamic login flow, better UX.
 😟A bit less secure by giving a hint that 2FA is required/wrong.
 
Personally, I go for the first `user:invalid_login_credential` because the user should usually know whether they have enabled 2FA or not and we already don't give any clue about `username` and `password` when they are wrong.

Reply:
I like this way :+1: The current flow is 2-step verification, but it feels a bit scary for me too because having to manipulate the authentication token.

I agree with passing the 2FA token to /admin.login. However, I have to choose the error response when the enabled-2fa user login with correct username/password but missing 2FA authentication token, what is a more appropriate error code between:

user:invalid_login_credentials:
:smiley:More secure by not giving a hint that which field is wrong.
:worried:A bit less dynamic login flow.

user:some_2fa_error:
:smiley:More dynamic login flow, better UX.
:worried:A bit less secure by giving a hint that 2FA is required/wrong.

Personally, I go for the first user:invalid_login_credential because the user should usually know whether they have enabled 2FA or not and we already don't give any clue about username and password when they are wrong.